File: sdk/reproit-android/src/main/kotlin/com/reproit/android/StructuralContracts.kt
package com.reproit.android

enum class ReproItContractStatus {
  VIOLATION,
  SATISFIED,
  ABSTAIN,
}

enum class ReproItStateBoundary(val wire: String) {
  ROTATION("rotation"),
  BACKGROUND_FOREGROUND("background-foreground"),
  NAVIGATION_ROUND_TRIP("navigation-round-trip"),
  PROCESS_RECREATION("process-recreation"),
}

enum class ReproItBoundaryPhase {
  BEFORE,
  AFTER,
}

data class ReproItStructuralObservation(
  val key: String,
  val state: String,
  val authoritative: Boolean,
  val settled: Boolean,
)

data class ReproItContractResult(
  val status: ReproItContractStatus,
  val id: String,
  val message: String? = null,
)

data class ReproItStatePreservationContract(
  val boundaries: Set<ReproItStateBoundary>,
  val sample: () -> ReproItStructuralObservation?,
  val saveBaseline: ((ReproItStateBoundary, ReproItStructuralObservation) -> Boolean)? = null,
  val loadBaseline: ((ReproItStateBoundary) -> ReproItStructuralObservation?)? = null,
)

internal object StatePreservationContracts {
  private val contracts = linkedMapOf<String, ReproItStatePreservationContract>()
  private val baselines = hashMapOf<String, ReproItStructuralObservation>()

  @Synchronized
  fun register(id: String, c: ReproItStatePreservationContract) {
    if (id.isNotEmpty() && c.boundaries.isNotEmpty()) contracts[id] = c
  }

  @Synchronized
  fun clear() {
    contracts.clear()
    baselines.clear()
  }

  @Synchronized
  fun boundary(
    kind: ReproItStateBoundary,
    phase: ReproItBoundaryPhase,
  ): List<ReproItContractResult> {
    val out = arrayListOf<ReproItContractResult>()
    for ((id, c) in contracts.toSortedMap()) {
      if (!c.boundaries.contains(kind)) continue
      val identity = "state-preservation:${kind.wire}:$id"
      val key = "${kind.wire}:$id"
      if (phase == ReproItBoundaryPhase.BEFORE) {
        val value = sampleState(c.sample)
        if (!valid(value)) {
          out += abstain(identity)
          continue
        }
        baselines[key] = value!!
        if (
          kind == ReproItStateBoundary.PROCESS_RECREATION &&
            (c.saveBaseline == null || safeBool { c.saveBaseline.invoke(kind, value) } != true)
        ) {
          baselines.remove(key)
          out += abstain(identity)
        } else out += satisfied(identity)
        continue
      }
      val before =
        if (kind == ReproItStateBoundary.PROCESS_RECREATION)
          c.loadBaseline?.let { sampleState { it(kind) } }
        else baselines[key]
      val after = sampleState(c.sample)
      baselines.remove(key)
      if (!valid(before) || !valid(after)) out += abstain(identity)
      else if (before!!.key == after!!.key && before.state == after.state) out += satisfied(identity)
      else
        out += violation(identity, "declared structural state was not preserved across ${kind.wire}")
    }
    return out
  }
}

data class ReproItActionEffectObservation(
  val route: String? = null,
  val state: String? = null,
  val authoritative: Boolean,
  val settled: Boolean,
)

data class ReproItTargetEffect(val target: String)

data class ReproItChangeEffect(val target: String? = null, val changed: Boolean? = null)

data class ReproItActionEffectContract(
  val sample: () -> ReproItActionEffectObservation?,
  val route: ReproItTargetEffect? = null,
  val state: ReproItChangeEffect? = null,
)

internal object ActionEffectContracts {
  private val contracts = linkedMapOf<String, ReproItActionEffectContract>()
  private val before = hashMapOf<String, ReproItActionEffectObservation>()

  @Synchronized
  fun register(id: String, c: ReproItActionEffectContract) {
    if (id.isNotEmpty()) contracts[id] = c
  }

  @Synchronized
  fun clear() {
    contracts.clear()
    before.clear()
  }

  @Synchronized
  fun begin(id: String): List<ReproItContractResult> {
    val c = contracts[id] ?: return listOf(abstain("action-effect:$id"))
    val value = sampleEffect(c.sample)
    if (!valid(value)) return listOf(abstain("action-effect:$id"))
    before[id] = value!!
    return listOf(satisfied("action-effect:$id"))
  }

  @Synchronized
  fun end(id: String): List<ReproItContractResult> {
    val c = contracts[id]
    val old = before.remove(id)
    val now = c?.let { sampleEffect(it.sample) }
    if (c == null || !valid(old) || !valid(now)) return listOf(abstain("action-effect:$id"))
    val out = arrayListOf<ReproItContractResult>()
    c.route?.let { expected -> checkTarget(out, id, "route", expected.target, now!!.route) }
    c.state?.let { expected -> checkChange(out, id, "state", expected, old!!.state, now!!.state) }
    return if (out.isEmpty()) listOf(abstain("action-effect:$id")) else out
  }
}

internal fun contractMarker(results: List<ReproItContractResult>): String? {
  val items =
    results
      .filter { it.status == ReproItContractStatus.VIOLATION }
      .map { mapOf("id" to it.id, "message" to (it.message ?: it.id)) }
  return if (items.isEmpty()) null
  else "REPROIT_INVARIANT " + Json.encode(mapOf("sig" to "", "items" to items))
}

private fun sampleState(f: () -> ReproItStructuralObservation?) =
  try {
    f()
  } catch (_: Throwable) {
    null
  }

private fun sampleEffect(f: () -> ReproItActionEffectObservation?) =
  try {
    f()
  } catch (_: Throwable) {
    null
  }

private fun valid(o: ReproItStructuralObservation?) =
  o != null && o.authoritative && o.settled && o.key.isNotEmpty() && o.state.isNotEmpty()

private fun valid(o: ReproItActionEffectObservation?) = o != null && o.authoritative && o.settled

private fun safeBool(f: () -> Boolean) =
  try {
    f()
  } catch (_: Throwable) {
    false
  }

private fun abstain(id: String) = ReproItContractResult(ReproItContractStatus.ABSTAIN, id)

private fun satisfied(id: String) = ReproItContractResult(ReproItContractStatus.SATISFIED, id)

private fun violation(id: String, message: String) =
  ReproItContractResult(ReproItContractStatus.VIOLATION, id, message)

private fun checkTarget(
  out: MutableList<ReproItContractResult>,
  id: String,
  kind: String,
  target: String,
  after: String?,
) {
  val identity = "action-effect:$id:$kind"
  out +=
    if (target.isEmpty() || after == null) abstain(identity)
    else if (target == after) satisfied(identity)
    else violation(identity, "declared $kind effect did not occur")
}

private fun checkChange(
  out: MutableList<ReproItContractResult>,
  id: String,
  kind: String,
  e: ReproItChangeEffect,
  before: String?,
  after: String?,
) {
  val identity = "action-effect:$id:$kind"
  if (after == null || (e.target == null && (e.changed == null || before == null))) {
    out += abstain(identity)
    return
  }
  val ok = if (e.target != null) after == e.target else (after != before) == e.changed
  out += if (ok) satisfied(identity) else violation(identity, "declared $kind effect did not occur")
}

File: sdk/reproit-ios/Sources/ReproIt/StructuralContracts.swift
import Foundation

public enum ReproItContractStatus: String {
  case violation = "VIOLATION"
  case satisfied = "SATISFIED"
  case abstain = "ABSTAIN"
}
public enum ReproItStateBoundary: String {
  case rotation
  case backgroundForeground = "background-foreground"
  case navigationRoundTrip = "navigation-round-trip"
  case processRecreation = "process-recreation"
}
public enum ReproItBoundaryPhase { case before, after }
public struct ReproItStructuralObservation {
  public let key: String, state: String, authoritative: Bool, settled: Bool
  public init(key: String, state: String, authoritative: Bool, settled: Bool) {
    self.key = key
    self.state = state
    self.authoritative = authoritative
    self.settled = settled
  }
}
public struct ReproItContractResult {
  public let status: ReproItContractStatus, id: String, message: String?
}
public struct ReproItStatePreservationContract {
  public let boundaries: Set<ReproItStateBoundary>, sample: () -> ReproItStructuralObservation?
  public let saveBaseline: ((ReproItStateBoundary, ReproItStructuralObservation) -> Bool)?
  public let loadBaseline: ((ReproItStateBoundary) -> ReproItStructuralObservation?)?
  public init(
    boundaries: Set<ReproItStateBoundary>, sample: @escaping () -> ReproItStructuralObservation?,
    saveBaseline: ((ReproItStateBoundary, ReproItStructuralObservation) -> Bool)? = nil,
    loadBaseline: ((ReproItStateBoundary) -> ReproItStructuralObservation?)? = nil
  ) {
    self.boundaries = boundaries
    self.sample = sample
    self.saveBaseline = saveBaseline
    self.loadBaseline = loadBaseline
  }
}
enum ReproItStatePreservationContracts {
  static var contracts: [String: ReproItStatePreservationContract] = [:],
    baselines: [String: ReproItStructuralObservation] = [:]
  static func register(_ id: String, _ c: ReproItStatePreservationContract) {
    if !id.isEmpty && !c.boundaries.isEmpty { contracts[id] = c }
  }
  static func clear() {
    contracts.removeAll()
    baselines.removeAll()
  }
  static func boundary(_ kind: ReproItStateBoundary, _ phase: ReproItBoundaryPhase)
    -> [ReproItContractResult]
  {
    var out: [ReproItContractResult] = []
    for (id, c) in contracts.sorted(by: { $0.key < $1.key }) where c.boundaries.contains(kind) {
      let identity = "state-preservation:\(kind.rawValue):\(id)"
      let key = "\(kind.rawValue):\(id)"
      if phase == .before {
        let value = sample(c.sample)
        guard valid(value) else {
          out.append(abstain(identity))
          continue
        }
        baselines[key] = value!
        if kind == .processRecreation
          && (c.saveBaseline == nil || safe { c.saveBaseline!(kind, value!) } != true)
        {
          baselines[key] = nil
          out.append(abstain(identity))
        } else {
          out.append(satisfiedResult(identity))
        }
        continue
      }
      let before =
        kind == .processRecreation
        ? (c.loadBaseline.flatMap { f in sample { f(kind) } }) : baselines[key]
      let after = sample(c.sample)
      baselines[key] = nil
      guard valid(before), valid(after) else {
        out.append(abstain(identity))
        continue
      }
      if before!.key == after!.key && before!.state == after!.state {
        out.append(satisfiedResult(identity))
      } else {
        out.append(
          violation(identity, "declared structural state was not preserved across \(kind.rawValue)"))
      }
    }
    return out
  }
}

public struct ReproItActionEffectObservation {
  public let route: String?, state: String?, authoritative: Bool, settled: Bool
  public init(route: String? = nil, state: String? = nil, authoritative: Bool, settled: Bool) {
    self.route = route
    self.state = state
    self.authoritative = authoritative
    self.settled = settled
  }
}
public struct ReproItTargetEffect {
  public let target: String
  public init(_ target: String) { self.target = target }
}
public struct ReproItChangeEffect {
  public let target: String?, changed: Bool?
  public init(target: String? = nil, changed: Bool? = nil) {
    self.target = target
    self.changed = changed
  }
}
public struct ReproItActionEffectContract {
  public let sample: () -> ReproItActionEffectObservation?, route: ReproItTargetEffect?,
    state: ReproItChangeEffect?
  public init(
    sample: @escaping () -> ReproItActionEffectObservation?, route: ReproItTargetEffect? = nil,
    state: ReproItChangeEffect? = nil
  ) {
    self.sample = sample
    self.route = route
    self.state = state
  }
}
enum ReproItActionEffectContracts {
  static var contracts: [String: ReproItActionEffectContract] = [:],
    before: [String: ReproItActionEffectObservation] = [:]
  static func register(_ id: String, _ c: ReproItActionEffectContract) {
    if !id.isEmpty { contracts[id] = c }
  }
  static func clear() {
    contracts.removeAll()
    before.removeAll()
  }
  static func begin(_ id: String) -> [ReproItContractResult] {
    guard let c = contracts[id], let value = sampleEffect(c.sample), valid(value) else {
      return [abstain("action-effect:\(id)")]
    }
    before[id] = value
    return [satisfiedResult("action-effect:\(id)")]
  }
  static func end(_ id: String) -> [ReproItContractResult] {
    guard let c = contracts[id], let old = before.removeValue(forKey: id),
      let now = sampleEffect(c.sample), valid(old), valid(now)
    else { return [abstain("action-effect:\(id)")] }
    var out: [ReproItContractResult] = []
    if let e = c.route { checkTarget(&out, id, "route", e.target, now.route) }
    if let e = c.state { checkChange(&out, id, "state", e, old.state, now.state) }
    return out.isEmpty ? [abstain("action-effect:\(id)")] : out
  }
}
func reproitContractMarker(_ results: [ReproItContractResult]) -> String? {
  let items = results.filter { $0.status == .violation }.map {
    ["id": $0.id, "message": $0.message ?? $0.id]
  }
  guard !items.isEmpty,
    let d = try? JSONSerialization.data(withJSONObject: ["sig": "", "items": items]),
    let j = String(data: d, encoding: .utf8)
  else { return nil }
  return "REPROIT_INVARIANT \(j)"
}
private func sample(_ f: () -> ReproItStructuralObservation?) -> ReproItStructuralObservation? {
  f()
}
private func sampleEffect(_ f: () -> ReproItActionEffectObservation?)
  -> ReproItActionEffectObservation?
{ f() }
private func safe(_ f: () -> Bool) -> Bool { f() }
private func valid(_ o: ReproItStructuralObservation?) -> Bool {
  o != nil && o!.authoritative && o!.settled && !o!.key.isEmpty && !o!.state.isEmpty
}
private func valid(_ o: ReproItActionEffectObservation?) -> Bool {
  o != nil && o!.authoritative && o!.settled
}
private func abstain(_ id: String) -> ReproItContractResult {
  .init(status: .abstain, id: id, message: nil)
}
private func satisfiedResult(_ id: String) -> ReproItContractResult {
  .init(status: .satisfied, id: id, message: nil)
}
private func violation(_ id: String, _ message: String) -> ReproItContractResult {
  .init(status: .violation, id: id, message: message)
}
private func checkTarget(
  _ out: inout [ReproItContractResult], _ id: String, _ kind: String, _ target: String,
  _ after: String?
) {
  let identity = "action-effect:\(id):\(kind)"
  guard !target.isEmpty, let after = after else {
    out.append(abstain(identity))
    return
  }
  out.append(
    after == target
      ? satisfiedResult(identity) : violation(identity, "declared \(kind) effect did not occur"))
}
private func checkChange(
  _ out: inout [ReproItContractResult], _ id: String, _ kind: String, _ e: ReproItChangeEffect,
  _ before: String?, _ after: String?
) {
  let identity = "action-effect:\(id):\(kind)"
  guard let after = after, e.target != nil || (e.changed != nil && before != nil) else {
    out.append(abstain(identity))
    return
  }
  let ok = e.target != nil ? after == e.target! : (after != before!) == e.changed!
  out.append(ok ? satisfiedResult(identity) : violation(identity, "declared \(kind) effect did not occur"))
}
